### PR TITLE
updating hyper-parameter search

### PR DESCRIPTION
**Pull request by jreps**
- updating hyper-parameter search to data.frame with metric, fold, value and then hyper-parameter values

### CODE REVIEW

**Review by egillax — COMMENTED** [review]
I had a few questions/suggestions (see comments). 

But there is one thing I think would be a nice time to address in this PR. That is how we calculate the AUC (or any other metric) for the CV performance. Currently it is done using the out-of-fold predictions concatenated. Most other packages do it by averaging the per-fold metric. These two approaches will give approximately the same results as long as the risk estimates are calibrated the same way across the folds. However if that is not the case, the current approach will give strange results. As in you might have per-fold performances of 0.72, 0.71, 0.72 but the overall CV performance will be 0.69 (or even lower). I'm not sure how this affects model selection but to be on the safe side I think it would be better to average over the folds. It also saves one call to the AUC function.

File: R/CyclopsModels.R [hunk 1]
```diff
@@ -167,7 +167,17 @@ fitCyclopsModel <- function(
     
     cvPerFold <-  unlist(lapply(modelTrained$cv, function(x){x$out_sample_auc}))
     if(length(cvPerFold)>0){
-      names(cvPerFold) <- paste0('fold_auc', 1:length(cvPerFold))
+      cvPerFold <- data.frame(
+        metric = 'AUC',
+        fold = 1:length(cvPerFold),
```
[the remark this answers is not in the document]
Comment (jreps) [reply R/CyclopsModels.R:172]: No reason - this was the original way, then later in time I added in the CV for the sklearn that is also used by the r classifier.  We can add for consistency if needed. I think the CV performance is recorded somewhere else though, so this may be redundant.

File: R/SklearnClassifier.R [hunk 2]
```diff
@@ -416,13 +416,29 @@ computeGridPerformance <- function(prediction, param, performanceFunct = 'comput
     }
   }
   
-  hyperSummary <- c(performanceFunct, performance, performanceFold, unlist(paramString))
-  names(hyperSummary) <- c(
-    'Metric', 
-    'cvPerformance', 
-    paste0('cvPerformanceFold',1:length(performanceFold)),
-    names(param)
+  #hyperSummary <- c(performanceFunct, performance, performanceFold, unlist(paramString))
```
[the remark this answers is not in the document]
Comment (jreps) [reply R/SklearnClassifier.R:419]: thumbs up

**Comment by jreps** [comment]
re the way we calculate CV performance.  I originally did the mean of the folds but I found this was problematic when the outcome is rare as each fold CV has a lot of uncertainty.  That's what prompted me to combine the folds' predictions and do the performance on the combination (the estimate is more stable although in theory each fold may have slightly different calibration which could mess things up).  Currently we save enough details to calculate both the aggregate prediction performance and the aggregate performance across folds, which I think we should keep.  We could give the users the option of whether to aggregate predictions and do performance or aggregate performance when picking the optimal hyper-parameters?  

In general we may want to make hyper-parameter selection more customizable: inputs 1) predictions per fold, 2) whether the minimize or maximize and 3) performance metric function and output is the selected hyper parameters. We can then compare different metrics to pick hyper-parameters :)

File: R/SklearnClassifier.R [hunk 3]
```diff
@@ -416,13 +416,29 @@ computeGridPerformance <- function(prediction, param, performanceFunct = 'comput
...
+  #  names(param)
+  #)
+  paramValues <- unlist(paramString)
+  names(paramValues) <- names(param)
+  
+  hyperSummary <- as.data.frame(
+    c(
+      data.frame(
+        metric = performanceFunct,
+        fold = c(0,1:length(performanceFold)),
```
[the remark this answers is not in the document]
Comment (jreps) [reply R/SklearnClassifier.R:433]: yeah - I did think 0 would not be clear.

File: R/CyclopsModels.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (egillax) [reply R/CyclopsModels.R:172]: ok then it's fine by me to keep it this way, for some reason I thought it had the CV performance but I see now it's not.

**Comment by egillax** [comment]
I think we might need to discuss the CV aggregation more later. I'm not convinced but want to hear more about how it was problematic before. 

But I don't think it should block this PR. Feel free to merge!

**Review by egillax — APPROVED** [review]
